crypto/crypto/src/noise.rs

```diff
@@ -0,0 +1,587 @@
+            .map_err(|_| NoiseError::Encrypt)?;
+        mix_hash(&mut h, &encrypted_payload);
+        msg.extend_from_slice(&encrypted_payload);
+
+        // return
+        let handshake_state = InitiatorHandshakeState { h, ck, e };
+        Ok((handshake_state, msg))
+    }
+
+    /// A client can call this to finalize a connection, after receiving an answer from a server.
+    pub fn finalize_connection(
+        &self,
+        handshake_state: InitiatorHandshakeState,
+        received_message: Vec<u8>,
+    ) -> Result<(Vec<u8>, NoiseSession), NoiseError> {
+        // checks
+        if received_message.len() > MAX_SIZE_NOISE_MSG {
+            return Err(NoiseError::ReceivedMsgTooLarge);
+        }
+        // retrieve handshake state
+        let InitiatorHandshakeState { mut h, mut ck, e } = handshake_state;
+
+        // <- e
+        let mut remote_ephemeral = [0u8; X25519_PUBLIC_KEY_LENGTH];
+        let mut cursor = Cursor::new(received_message);
+        cursor
+            .read_exact(&mut remote_ephemeral)
+            .map_err(|_| NoiseError::MsgTooShort)?;
+        mix_hash(&mut h, &remote_ephemeral);
+        let remote_ephemeral = x25519::PublicKey::from(remote_ephemeral);
+
+        // <- ee
+        let dh_output = e.diffie_hellman(&remote_ephemeral);
+        mix_key(&mut ck, dh_output.as_bytes())?;
+
+        // <- se
+        let dh_output = self.private_key.diffie_hellman(&remote_ephemeral);
+        let k = mix_key(&mut ck, dh_output.as_bytes())?;
+
+        // <- payload
+        let received_encrypted_payload: Vec<_> = cursor.bytes().map(|x| x.unwrap()).collect();
+
+        #[cfg(test)]
+        let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&k));
+        #[cfg(not(test))]
+        let aead = Aes128Gcm::new(GenericArray::clone_from_slice(&k));
+
+        let nonce = GenericArray::from_slice(&[0u8; 12]);
+        let ct_and_ad = Payload {
+            msg: &received_encrypted_payload,
+            aad: &h,
+        };
+        let received_payload = match aead.decrypt(nonce, ct_and_ad) {
+            Ok(res) => res,
+            Err(_) if cfg!(feature = "fuzzing") => Vec::new(),
+            Err(_) => {
+                return Err(NoiseError::Decrypt);
+            }
+        };
+
+        // split
+        let (mut k1, mut k2) = hkdf(&ck, None)?;
+        if !cfg!(test) {
+            k1.resize(16, 0);
+            k2.resize(16, 0);
+        }
+        let session = NoiseSession::new(k1, k2);
+
+        //
+        Ok((received_payload, session))
+    }
+
+    //
+    // Responder
+    // ---------
+
+    /// The sole function the server can call to parse and respond to a client initiating a connection.
+    pub fn respond_to_client_and_finalize(
+        &self,
+        rng: &mut (impl rand::RngCore + rand::CryptoRng),
+        prologue: &[u8],
+        received_message: Vec<u8>,
+        payload: Option<&[u8]>,
+    ) -> Result<(Vec<u8>, x25519::PublicKey, Vec<u8>, NoiseSession), NoiseError> {
```

Comment:
this should have a `msg_buf: &mut [u8]` and `recv_payload_buf: &mut [u8]`

Reply:
why are these mut?

Reply:
you mean `received_message` and `payload` should be `&mut [u8]` right?

Reply:
i meant the caller passes in an output buffer for the bytes to send and the decrypted payload from the sender. anyway, we can investigate this later i guess.